Defer setting up enhanced recording until after PAM has completed initializing

rosstimothy opened the pull request:
The reexec process now has a two way wait mechanism to allow the child process to complete any setup operations that may be required before the parent process starts enhanced recording.

The old process was:

1) Parent launches child process
2) Child process opens PAM context and then waits for the continue signal before executing the command
3) Parent sets up enhanced recording
4) Parent sends the continue signal
5) Child executes command/opens shell


The new process is now:

1) Parent launches child process and waits for child continue signal
2) Child process opens PAM context and then signals it has completed setup
3) Parent receives child continue signal and sets up enhanced recording
4) Parent sends the continue signal
5) Child executes command/opens shell

Closes https://github.com/gravitational/teleport/issues/29030

jakule commented on lib/srv/reexec.go:
```diff
@@ -317,6 +338,18 @@ func RunCommand() (errw io.Writer, code int, err error) {
+	// and that we are now waiting for the continue signal before proceeding. This is needed
+	// to ensure that PAM changing the cgroup doesn't bypass enhanced recording.
+	ready = true
+	if _, err := readyfd.Write([]byte{0}); err != nil {
```
Since this is a different fd, do we need to even write anything to it, or can we just close it to unblock? Value is being ignored after reading anyways.

jakule commented on lib/srv/reexec.go:
```diff
@@ -213,6 +217,23 @@ func RunCommand() (errw io.Writer, code int, err error) {
 	}
+	readyfd := os.NewFile(ReadyFile, fdName(ReadyFile))
+	if readyfd == nil {
+		return errorWriter, teleport.RemoteCommandFailure, trace.BadParameter("continue write pipe not found")
```
I checked that, and if the ready fd is not closed, the parent process blocks forever. If this error or any other above is returned, the parent blocks. TBH I'm not sure how we can fix it. We could add a timeout, but I'm not sure how long the PAM can take if you use some network based plugins.

jakule reviewed (approved):
LGTM. One nit is that we implicitly assume that PAM won't take more than 10 seconds. I hope this is a reasonable assumption.

rosstimothy committed:
Defer setting up enhanced recording until after PAM has completed

The reexec process now has a two way wait mechanism to allow the
child process to complete any setup operations that may be required
before the parent process starts enhanced recording.

The old process was:

1) Parent launches child process
2) Child process opens PAM context and blocks on the continue signal
3) Parent sets up enhanced recording
4) Parent sends the continue signal
5) Child executes command/opens shell

The new process is:

1) Parent launches child process and waits for child continue signal
2) Child process opens PAM context and then signals it has completed setup
3) Parent receives child continue signal and sets up enhanced recording
4) Parent sends the continue signal
5) Child executes command/opens shell

Closes #29030
